### ISSUE TITLE
Prebuilt libcurl/7.56.1 for Windows doesn't seem to have OpenSSL support

### ISSUE THREAD

**Issue by hdk5** [bincrafters/community#192]
### Package Details
* Package Name/Version: **libcurl/7.56.1**
* Operating System: **Windows**
* Compiler+version: **Visual Studio 15**

### Steps to reproduce
This is supported protocols I get from curl_version_info using libcurl/7.56.1 package:
> dict file ftp gopher http imap pop3 rtsp smtp telnet tftp

No https there

And this is what I get from libcurl/7.52.1:
> dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smtp smtps telnet tftp

https and other secure protocols are here

**Comment by theirix** [comment]
Yes, we don't prebuild openssl-enabled libcurl on CI because builds are very heavy. You can always build it on your machine by specifying an option `-o libcurl:with_openssl=True` on command line on in your `conanfile.txt`.

**Comment by hdk5** [comment]
I have this option in my `conanfile.txt`, but it tries to use prebuilt package unless I say him to `--build libcurl` himself. 

Shouldn't it throw error like
> ERROR: Missing prebuilt package for 'libcurl/7.56.1@bincrafters/stable'```

instead?

**Comment by theirix** [comment]
Just add `--build=missing`. So your command line will be like `conan install --build=missing .`

**Comment by hdk5** [comment]
The thing is conan does NOT throw this error, but should.

**Comment by theirix** [comment]
Hmm, you are right, this looks strange. Could you please attach your `conanfile.txt` and the output of `conan info .`?

**Comment by hdk5** [comment]
```
hdk5@HDK5-LAPTOP C:\Users\hdk5\repo\_d1\build
> type ..\conanfile.txt
[requires]
libcurl/7.56.1@bincrafters/stable

[generators]
cmake

[options]
libcurl:with_openssl=True

hdk5@HDK5-LAPTOP C:\Users\hdk5\repo\_d1\build
> conan info ..
PROJECT
    ID: 00026ab163077014d4d56f7925680cb99e5b898c
    BuildID: None
    Requires:
        libcurl/7.56.1@bincrafters/stable
OpenSSL/1.0.2n@conan/stable
    ID: 606fdb601e335c2001bdf31d478826b644747077
    BuildID: None
    Remote: conan-center=https://conan.bintray.com
    URL: http://github.com/lasote/conan-openssl
    License: The current OpenSSL licence is an 'Apache style' license: https://www.openssl.org/source/license.html
    Updates: Version not checked
    Creation date: 2018-03-15 13:37:39
    Required by:
        libcurl/7.56.1@bincrafters/stable
    Requires:
        zlib/1.2.11@conan/stable
libcurl/7.56.1@bincrafters/stable
    ID: ae6191bc1194317ff7cdbc84d4d302ab49975618
    BuildID: None
    Remote: bincrafters=https://api.bintray.com/conan/bincrafters/public-conan
    URL: http://github.com/bincrafters/conan-libcurl
    License: MIT
    Updates: Version not checked
    Creation date: 2018-03-16 00:14:57
    Required by:
        PROJECT
    Requires:
        OpenSSL/1.0.2n@conan/stable
        zlib/1.2.11@conan/stable
zlib/1.2.11@conan/stable
    ID: 6cc50b139b9c3d27b3e9042d5f5372d327b3a9f7
    BuildID: None
    Remote: conan-center=https://conan.bintray.com
    URL: http://github.com/lasote/conan-zlib
    License: http://www.zlib.net/zlib_license.html
    Updates: Version not checked
    Creation date: 2018-03-06 12:57:33
    Required by:
        OpenSSL/1.0.2n@conan/stable
        libcurl/7.56.1@bincrafters/stable
```

**Comment by theirix** [comment]
By the way, libcurl built on Windows WITH openssl does not enable openssl nor winssl. I'll try to fix it later.

But we should ensure that Conan picks the right package. Package ae6191bc1194317ff7cdbc84d4d302ab49975618 is built with openssl. It can be verified by your conan info or by looking into its conaninfo.txt.

Could you please post which libcurl packages do you have in your local cache by running
`conan search libcurl/7.56.1@bincrafters/stable` ? I expect you have a few libcurl packages with or without openssl.

To reproduce:
1. remove libcurl with openssl (ae6191bc1194317ff7cdbc84d4d302ab49975618) from your local cache:
`conan remove -p ae6191bc1194317ff7cdbc84d4d302ab49975618 libcurl/7.56.1@bincrafters/stable`
or even remove all libcurl packages.

2. build again without and `--build` switches: `conan install .`. Conan must throw a `missing prebuilt` exception. If it does not, it is a conan bug.

Additional tips:
- run `conan install --update .` to have latest versions
- are you sure you are using 7.56.1 or 7.52.1 in your original post? they may differ a little.

**Comment by theirix** [comment]
> It still installs libcurl/7.56.1@bincrafters/stable:ae6191bc1194317ff7cdbc84d4d302ab49975618 package, which, as you said, was built with openssl, but does not support https protocol.

> When I have libcurl/7.52.1@bincrafters/stable in conanfile.txt, it installs libcurl/7.52.1@bincrafters/stable:ae6191bc1194317ff7cdbc84d4d302ab49975618 package, which supports https protocol.

So it means that 7.56.1 with with_openssl=True does not actually supports openssl.
And the 7.52.1 with with_openssl=True supports it.
It's an ongoing problem and I'll track it in this ticket.

> Is it ok that 7.52.1 and 7.56.1 have the same ae619... hash?

Yes, it's ok. Package id is a hash of settings and options but not the version. Package is identified  by name, version, username, channel and package id.


Regarding the `missing` problem. You can try to use 7.56.1+with_openssl=False by changing conanfile and using `--build=missing`. Then try to use  7.56.1+with_openssl=True and do NOT use `--build`. I think you had a prebuilt package in your local cache and it was picked up without prebuilding.

**Comment by theirix** [comment]
Fixed all libcurl versions to enable openssl. Also added an option `with_winssl` (default off) to build with windows-native SSL backend, it will enable https too.

**Comment by solvingj** [comment]
great work on this, it was a tremendous amount of work and annoyance, thank you for doing it.
